hadoop-hdds/common/src/main/java/org/apache/hadoop/ozone/common/IncrementalChunkBuffer.java

```diff
@@ -0,0 +1,266 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ *  with the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+package org.apache.hadoop.ozone.common;
+
+import com.google.common.base.Preconditions;
+import org.apache.ratis.thirdparty.com.google.protobuf.ByteString;
+
+import java.nio.BufferOverflowException;
+import java.nio.ByteBuffer;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Optional;
+import java.util.function.Function;
+
+/**
+ * Use a list of {@link ByteBuffer} to implement a single {@link ChunkBuffer}
+ * so that the buffer can be allocated incrementally.
+ */
+final class IncrementalChunkBuffer implements ChunkBuffer {
+  /** Limit, which cannot be changed. */
```

Comment:
what is the definition of Limit, is it similar to a byteBuffer in Java ? if you have list of buffers, each of the buffers defined by increment, what is the role of limit here ?

Reply:
Limit is the limit of the entire ChunkBuffer, not individual ByteBuffers.  When we refactor the code from ByteBuffer to ChunkBuffer, we preserve the use "limit" in order to minimize the changes.